File: mobile/modules/vault-shared/android/src/main/java/expo/modules/vaultshared/VaultSharedModule.kt
package expo.modules.vaultshared

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class VaultSharedModule : Module() {
  private var handle: Long = 0

  override fun definition() = ModuleDefinition {
    Name("VaultShared")

    AsyncFunction("initialize") { basePath: String, dataPath: String ->
      destroyHandle()

      val nextHandle = nativeCreate(basePath, dataPath)
      if (nextHandle == 0L) {
        throw RuntimeException(nativeTakeLastError())
      }

      handle = nextHandle
    }

    AsyncFunction("waitForScan") { timeoutMs: Long ->
      if (handle == 0L) false else nativeWaitForScan(handle, timeoutMs)
    }

    AsyncFunction("getProgressJson") {
      nativeProgressJson(requireHandle())
    }

    AsyncFunction("searchFilesJson") { query: String, limit: Long ->
      nativeSearchFilesJson(requireHandle(), query, limit)
    }

    AsyncFunction("noteSearchJson") { query: String, scope: String ->
      nativeNoteSearchJson(requireHandle(), query, scope)
    }

    AsyncFunction("searchTrackSelectionJson") { query: String, notePath: String ->
      nativeSearchTrackSelectionJson(requireHandle(), query, notePath)
    }

    Function("dispose") {
      destroyHandle()
    }
  }

  private fun requireHandle(): Long {
    if (handle == 0L) {
      throw RuntimeException("VaultShared is not initialized")
    }

    return handle
  }

  private fun destroyHandle() {
    if (handle != 0L) {
      nativeDestroy(handle)
      handle = 0
    }
  }

  private external fun nativeCreate(basePath: String, dataPath: String): Long
  private external fun nativeDestroy(handle: Long)
  private external fun nativeWaitForScan(handle: Long, timeoutMs: Long): Boolean
  private external fun nativeProgressJson(handle: Long): String
  private external fun nativeSearchFilesJson(handle: Long, query: String, limit: Long): String
  private external fun nativeNoteSearchJson(handle: Long, query: String, scope: String): String
  private external fun nativeSearchTrackSelectionJson(
    handle: Long,
    query: String,
    notePath: String,
  ): String
  private external fun nativeTakeLastError(): String

  companion object {
    init {
      System.loadLibrary("vault_shared_ffi")
    }
  }
}

File: mobile/modules/vault-shared/ios/VaultSharedModule.swift
import ExpoModulesCore

@_silgen_name("vault_shared_search_create")
private func vaultSharedSearchCreate(
  _ basePath: UnsafePointer<CChar>,
  _ dataPath: UnsafePointer<CChar>
) -> UnsafeMutableRawPointer?

@_silgen_name("vault_shared_search_destroy")
private func vaultSharedSearchDestroy(_ handle: UnsafeMutableRawPointer?)

@_silgen_name("vault_shared_search_wait_for_scan")
private func vaultSharedSearchWaitForScan(_ handle: UnsafeMutableRawPointer?, _ timeoutMs: UInt64) -> Bool

@_silgen_name("vault_shared_search_progress_json")
private func vaultSharedSearchProgressJson(_ handle: UnsafeMutableRawPointer?) -> UnsafeMutablePointer<CChar>?

@_silgen_name("vault_shared_search_files_json")
private func vaultSharedSearchFilesJson(
  _ handle: UnsafeMutableRawPointer?,
  _ query: UnsafePointer<CChar>,
  _ limit: UInt32
) -> UnsafeMutablePointer<CChar>?

@_silgen_name("vault_shared_free_string")
private func vaultSharedFreeString(_ value: UnsafeMutablePointer<CChar>?)

@_silgen_name("vault_shared_take_last_error")
private func vaultSharedTakeLastError() -> UnsafeMutablePointer<CChar>?

public class VaultSharedModule: Module {
  private var handle: UnsafeMutableRawPointer?

  deinit {
    destroyHandle()
  }

  public func definition() -> ModuleDefinition {
    Name("VaultShared")

    AsyncFunction("initialize") { (basePath: String, dataPath: String) in
      self.destroyHandle()

      let nextHandle = basePath.withCString { basePathPointer in
        dataPath.withCString { dataPathPointer in
          vaultSharedSearchCreate(basePathPointer, dataPathPointer)
        }
      }

      guard let nextHandle else {
        throw NSError(
          domain: "VaultShared",
          code: 1,
          userInfo: [NSLocalizedDescriptionKey: self.takeLastError()]
        )
      }

      self.handle = nextHandle
    }

    AsyncFunction("waitForScan") { (timeoutMs: UInt64) in
      guard let handle = self.handle else {
        return false
      }

      return vaultSharedSearchWaitForScan(handle, timeoutMs)
    }

    AsyncFunction("getProgressJson") {
      try self.takeJson(vaultSharedSearchProgressJson(try self.requireHandle()))
    }

    AsyncFunction("searchFilesJson") { (query: String, limit: UInt32) in
      try query.withCString { queryPointer in
        try self.takeJson(vaultSharedSearchFilesJson(try self.requireHandle(), queryPointer, limit))
      }
    }

    Function("dispose") {
      self.destroyHandle()
    }
  }

  private func requireHandle() throws -> UnsafeMutableRawPointer {
    guard let handle else {
      throw NSError(
        domain: "VaultShared",
        code: 2,
        userInfo: [NSLocalizedDescriptionKey: "VaultShared is not initialized"]
      )
    }

    return handle
  }

  private func destroyHandle() {
    if let handle {
      vaultSharedSearchDestroy(handle)
      self.handle = nil
    }
  }

  private func takeJson(_ pointer: UnsafeMutablePointer<CChar>?) throws -> String {
    guard let pointer else {
      throw NSError(
        domain: "VaultShared",
        code: 3,
        userInfo: [NSLocalizedDescriptionKey: "VaultShared returned an empty response"]
      )
    }

    defer {
      vaultSharedFreeString(pointer)
    }

    return String(cString: pointer)
  }

  private func takeLastError() -> String {
    guard let pointer = vaultSharedTakeLastError() else {
      return "Unknown VaultShared native error"
    }

    defer {
      vaultSharedFreeString(pointer)
    }

    return String(cString: pointer)
  }
}
